prow/scheduler/strategy/roundrobin.go

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategy
+
+import (
+	"context"
+
+	prowv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+// RoundRobin is a scheduling strategy that implements round-robin scheduling.
+// The list of clusters to iterate over is given by prow-controller-manager.
+type RoundRobin struct {
```

Comment:
Instead of passing a list of clusters and another variable to indicate the index of the clusters slice, it would be better to implement a queue instead.

Reply:
The reconciliation process is happening concurrently. We also need to take care of the race conditions here. Impelementing a queue should be safer. You won't need an index variable since you always want to move the index once in the array.

Reply:
It looks a bit an overkill as the round robin I seek is really no more than two lines long:
```golang
i := rr.i
rr.i = (rr.i + 1) % uint(len(rr.clusters))
```
and even the thread safe variant is quite simple:
```golang
type RoundRobin struct {
  m sync.Mutex
  // ...
}

func (rr *RoundRobin) Schedule(_ context.Context, _ *prowv1.ProwJob) (Result, error) {
  m.Lock()
  defer m.Unlock()
  i := rr.i
  rr.i = (rr.i + 1) % uint(len(rr.clusters))
  // ...
}

```
We are trading a `i uint` for an entire new type, and that means more code and more tests to accomplish. Furthermore `append` may reallocate the slice (not going into the internals here, too much of a stretch).
I'd agree to introduce a `Queue` whenever we need round robin in more than a single place, but here we only need something that simple.

Reply:
Also, how prow will know which of those cluster are healthy and why do we need to failover?

Reply:
A round-robin strategy doesn't make sense to run in parallel because you have to - by definition - sequentially pop the next target from the head. The current implementation seems to be sufficient.